src/components/mc/cuda/mc_cuda.c

```diff
@@ -212,50 +224,172 @@ static ucc_status_t ucc_mc_cuda_init()
+    // spinlock assures 1 thread only initiates mpool and allocates its buffers
+    ucc_spinlock_init(&ucc_mc_cuda.mpool_init_spinlock, 0);
 
     return UCC_OK;
 }
 
 static ucc_status_t ucc_mc_cuda_finalize()
 {
     CUDACHECK(cudaStreamDestroy(ucc_mc_cuda.stream));
+    if (ucc_mc_cuda.mpool_init_flag) {
+        ucc_mpool_cleanup(&ucc_mc_cuda.mpool, 1);
+        ucc_mc_cuda.mpool_init_flag     = 0;
+        ucc_mc_cuda.super.ops.mem_alloc = ucc_mc_cuda_mem_alloc_pool_with_init;
+    }
+    ucc_spinlock_destroy(&ucc_mc_cuda.mpool_init_spinlock);
     return UCC_OK;
 }
 
-static ucc_status_t ucc_mc_cuda_mem_alloc(void **ptr, size_t size)
+static ucc_status_t
+ucc_mc_cuda_mem_alloc(ucc_mc_buffer_header_t **ptr, size_t size)
 {
-    cudaError_t st;
-
-    st = cudaMalloc(ptr, size);
+    cudaError_t             st;
+    ucc_mc_buffer_header_t *h =
+        ucc_malloc(sizeof(ucc_mc_buffer_header_t), "mc cuda");
+    if (ucc_unlikely(!h)) {
+        mc_error(&ucc_mc_cuda.super, "failed to allocate %zd bytes",
+                 sizeof(ucc_mc_buffer_header_t));
+        return UCC_ERR_NO_MEMORY;
+    }
+    st = cudaMalloc(&h->addr, size);
     if (ucc_unlikely(st != cudaSuccess)) {
         cudaGetLastError();
         mc_error(&ucc_mc_cuda.super,
                  "failed to allocate %zd bytes, "
                  "cuda error %d(%s)",
                  size, st, cudaGetErrorString(st));
+        ucc_free(h);
+        return UCC_ERR_NO_MEMORY;
+    }
+    mc_debug(
+        &ucc_mc_cuda.super, "ucc_mc_cuda_mem_alloc size:%zd",
+        size); // TODO: should i keep this? add it also to chunk alloc? in cpu?
+    h->from_pool = 0;
+    *ptr         = h;
+    return UCC_OK;
+}
+
+static ucc_status_t ucc_mc_cuda_mem_alloc_pool(ucc_mc_buffer_header_t **ptr,
+                                          size_t                   size)
+{
+    ucc_mc_buffer_header_t *h = NULL;
+    if (size <= MC_CUDA_CONFIG->cuda_elem_size) {
+        h = (ucc_mc_buffer_header_t *)ucc_mpool_get(&ucc_mc_cuda.mpool);
+    }
+    if (!h) {
+        // Slow path
+    	return ucc_mc_cuda_mem_alloc(ptr, size);
+    }
+    *ptr = h;
+    return h->addr ? UCC_OK : UCC_ERR_NO_MEMORY;
+}
+
+static ucc_status_t ucc_mc_cuda_chunk_alloc(ucc_mpool_t *mp, size_t *size_p,
+                                            void **chunk_p)
+{
+    *chunk_p = ucc_malloc(*size_p, "mc cuda");
+    if (!*chunk_p) {
+        mc_error(&ucc_mc_cuda.super, "failed to allocate %zd bytes", *size_p);
         return UCC_ERR_NO_MEMORY;
     }
 
-    mc_debug(&ucc_mc_cuda.super, "ucc_mc_cuda_mem_alloc size:%ld", size);
     return UCC_OK;
 }
 
-static ucc_status_t ucc_mc_cuda_mem_free(void *ptr)
+static void ucc_mc_cuda_chunk_init(ucc_mpool_t *mp, void *obj, void *chunk)
 {
-    cudaError_t st;
+    ucc_mc_buffer_header_t *h = (ucc_mc_buffer_header_t *)obj;
+    cudaError_t st = cudaMalloc(&h->addr, MC_CUDA_CONFIG->cuda_elem_size);
+    if (st != cudaSuccess) {
+        cudaGetLastError();
```

Comment:
please add here documentation regarding how you check if h->addr is null in another function